File: i2pchat/sam/protocol.py
from __future__ import annotations

from dataclasses import dataclass
import re

from .errors import ProtocolError, map_result_to_error


SENSITIVE_SAM_KEYS = {"PRIV", "PRIVATE", "DESTINATION", "SIGNING_PRIVATE_KEY"}
TRANSIENT_DESTINATION = "TRANSIENT"


@dataclass(frozen=True, slots=True)
class SAMReply:
    command: str
    topic: str
    fields: dict[str, str]
    raw_line: str


def _validate_sam_token(value: object, *, field_name: str, allow_equals: bool) -> str:
    token = str(value or "").strip()
    if not token:
        raise ValueError(f"SAM {field_name} is empty")
    if any(ch in token for ch in ("\r", "\n", "\x00", " ", "\t")):
        raise ValueError(f"SAM {field_name} contains forbidden characters")
    if not allow_equals and "=" in token:
        raise ValueError(f"SAM {field_name} contains forbidden characters")
    return token


def _validate_session_id(session_id: object) -> str:
    return _validate_sam_token(session_id, field_name="ID", allow_equals=False)


def _validate_boolish_flag(value: object, *, field_name: str) -> str:
    token = _validate_sam_token(value, field_name=field_name, allow_equals=False).lower()
    if token not in {"true", "false"}:
        raise ValueError(f"SAM {field_name} is invalid")
    return token


def _validate_port(port: object) -> int:
    if type(port) is not int:
        raise ValueError("SAM PORT must be an int")
    if port < 1 or port > 65535:
        raise ValueError("SAM PORT is out of range")
    return port


def _validate_sam_options(options: object) -> str:
    text = str(options or "").strip()
    if any(ch in text for ch in ("\r", "\n", "\x00")):
        raise ValueError("SAM OPTIONS contains forbidden characters")
    return text


def _session_options_dict_to_tokens(
    options: dict[str, str],
    *,
    sig_type: int | None,
) -> str:
    """Space-joined name=value tokens for SESSION CREATE (after DESTINATION)."""
    bits: list[str] = []
    if sig_type is not None:
        bits.append(f"SIGNATURE_TYPE={int(sig_type)}")
    if options:
        for raw_key, raw_val in options.items():
            for label, raw in (("OPTION_KEY", raw_key), ("OPTION_VALUE", raw_val)):
                s = str(raw or "")
                if any(ch in s for ch in ("\r", "\n", "\x00")):
                    raise ValueError(f"SAM {label} contains forbidden characters")
            key_s = _validate_sam_token(
                raw_key, field_name="OPTION_KEY", allow_equals=False
            )
            val_s = _validate_sam_token(
                raw_val, field_name="OPTION_VALUE", allow_equals=False
            )
            bits.append(f"{key_s}={val_s}")
    return " ".join(bits)


def _validate_sam_style(style: object) -> str:
    token = _validate_sam_token(style, field_name="STYLE", allow_equals=False).upper()
    if token not in {"STREAM", "DATAGRAM", "RAW"}:
        raise ValueError("SAM STYLE is invalid")
    return token


def _validate_sam_destination_token(destination: object) -> str:
    token = _validate_sam_token(
        destination, field_name="DESTINATION", allow_equals=True
    )
    if token == TRANSIENT_DESTINATION:
        return token
    return token


def _validate_sam_version(version: object, *, field_name: str) -> str:
    token = _validate_sam_token(version, field_name=field_name, allow_equals=False)
    if not re.fullmatch(r"\d+\.\d+", token):
        raise ValueError(f"SAM {field_name} version is invalid")
    return token


def build_hello(min_version: str = "3.0", max_version: str = "3.2") -> bytes:
    min_v = _validate_sam_version(min_version, field_name="MIN")
    max_v = _validate_sam_version(max_version, field_name="MAX")
    return f"HELLO VERSION MIN={min_v} MAX={max_v}\n".encode("ascii")


def build_dest_generate(sig_type: int = 7) -> bytes:
    if type(sig_type) is not int:
        raise ValueError("SAM SIGNATURE_TYPE must be an int")
    return f"DEST GENERATE SIGNATURE_TYPE={sig_type}\n".encode("ascii")


def build_naming_lookup(name: str) -> bytes:
    safe_name = _validate_sam_token(name, field_name="NAME", allow_equals=False)
    return f"NAMING LOOKUP NAME={safe_name}\n".encode("ascii")


def build_session_create(
    style: str,
    session_id: str,
    destination: str,
    option_string: str = "",
    *,
    sig_type: int | None = None,
    options: dict[str, str] | None = None,
) -> bytes:
    if option_string:
        if sig_type is not None or options:
            raise ValueError(
                "option_string is mutually exclusive with sig_type/options"
            )
        style_s = _validate_sam_style(style)
        session_s = _validate_session_id(session_id)
        dest_s = _validate_sam_destination_token(destination)
        opts_s = _validate_sam_options(option_string)
        return (
            f"SESSION CREATE STYLE={style_s} ID={session_s} "
            f"DESTINATION={dest_s} {opts_s}\n"
        ).encode("ascii")
    # SAM v3: I2CP/streaming options are plain name=value tokens after DESTINATION
    # (see geti2p SAM spec "[option=value]*"). Do not insert a standalone "OPTION"
    # keyword — SAM routers expect space-separated name=value pairs only.
    option_string = _session_options_dict_to_tokens(options or {}, sig_type=sig_type)
    style_s = _validate_sam_style(style)
    session_s = _validate_session_id(session_id)
    dest_s = _validate_sam_destination_token(destination)
    opts_s = _validate_sam_options(option_string)
    return (
        f"SESSION CREATE STYLE={style_s} ID={session_s} "
        f"DESTINATION={dest_s} {opts_s}\n"
    ).encode("ascii")


def build_stream_connect(
    session_id: str,
    destination: str,
    *,
    silent: str = "false",
) -> bytes:
    sid = _validate_session_id(session_id)
    dest = _validate_sam_token(
        destination, field_name="DESTINATION", allow_equals=True
    )
    silent_flag = _validate_boolish_flag(silent, field_name="SILENT")
    return (
        f"STREAM CONNECT ID={sid} DESTINATION={dest} SILENT={silent_flag}\n"
    ).encode("ascii")


def build_stream_accept(session_id: str) -> bytes:
    sid = _validate_session_id(session_id)
    return f"STREAM ACCEPT ID={sid} SILENT=false\n".encode("ascii")


def build_stream_forward(session_id: str, port: int) -> bytes:
    sid = _validate_session_id(session_id)
    safe_port = _validate_port(port)
    return f"STREAM FORWARD ID={sid} PORT={safe_port} \n".encode("ascii")


def parse_reply_line(line: bytes) -> SAMReply:
    raw = line.decode("utf-8", errors="replace").strip()
    if not raw:
        raise ProtocolError(message="Empty SAM reply", raw_line="")
    parts = raw.split()
    if len(parts) < 2:
        raise ProtocolError(message="Malformed SAM reply", raw_line=raw)
    command = parts[0].upper()
    topic = parts[1].upper()
    fields: dict[str, str] = {}
    for token in parts[2:]:
        if "=" not in token:
            continue
        key, value = token.split("=", 1)
        fields[key.upper()] = value
    return SAMReply(command=command, topic=topic, fields=fields, raw_line=raw)


def expect_ok(reply: SAMReply, *, result_key: str = "RESULT") -> SAMReply:
    result = reply.fields.get(result_key, "").upper()
    if result == "OK":
        return reply
    if not result:
        # i2pd (SAM 3.x) often omits RESULT=OK on successful DEST GENERATE: a single line
        # DEST REPLY PUB=… PRIV=… with no RESULT token. Treat PUB+PRIV as implicit OK.
        if (
            reply.command == "DEST"
            and reply.topic == "REPLY"
            and reply.fields.get("PUB")
            and reply.fields.get("PRIV")
        ):
            return reply
        # i2pd: successful SESSION CREATE may reply SESSION STATUS DESTINATION=… without RESULT=OK.
        if (
            reply.command == "SESSION"
            and reply.topic == "STATUS"
            and reply.fields.get("DESTINATION")
        ):
            return reply
        raise ProtocolError(message="SAM reply missing RESULT", raw_line=reply.raw_line)
    message = reply.fields.get("MESSAGE") or f"{reply.command} {reply.topic} failed"
    raise map_result_to_error(result, message, raw_line=reply.raw_line)


def _redact_sam_reply(raw_reply: str) -> str:
    if not raw_reply:
        return ""
    redacted_parts = []
    for token in raw_reply.split(" "):
        if "=" not in token:
            redacted_parts.append(token)
            continue
        key, value = token.split("=", 1)
        if key in SENSITIVE_SAM_KEYS and value:
            redacted_parts.append(f"{key}=<redacted>")
        else:
            redacted_parts.append(token)
    return " ".join(redacted_parts)


# Transition aliases for tests and incremental call-site migration.
hello = build_hello
dest_generate = build_dest_generate
naming_lookup = build_naming_lookup
session_create = build_session_create
stream_connect = build_stream_connect
stream_accept = build_stream_accept
stream_forward = build_stream_forward

File: cpp/core/src/sam/protocol.cpp
#include "i2pchat/sam/protocol.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <regex>
#include <set>
#include <sstream>

namespace i2pchat::sam {
namespace {

std::string to_upper(std::string_view text) {
    std::string out(text);
    std::transform(out.begin(), out.end(), out.begin(), [](unsigned char ch) {
        return static_cast<char>(std::toupper(ch));
    });
    return out;
}

std::string trim(std::string_view text) {
    const auto is_space = [](unsigned char ch) { return std::isspace(ch) != 0; };
    std::size_t begin = 0;
    while (begin < text.size() && is_space(static_cast<unsigned char>(text[begin]))) {
        ++begin;
    }
    std::size_t end = text.size();
    while (end > begin && is_space(static_cast<unsigned char>(text[end - 1]))) {
        --end;
    }
    return std::string(text.substr(begin, end - begin));
}

/// Reject characters that would let a caller inject extra SAM commands.
std::string validate_token(std::string_view value, std::string_view field_name,
                           bool allow_equals) {
    const std::string token = trim(value);
    if (token.empty()) {
        throw SamError(SamErrorKind::Protocol,
                       "SAM " + std::string(field_name) + " is empty", "");
    }
    static constexpr std::array<char, 5> kForbidden = {'\r', '\n', '\0', ' ', '\t'};
    for (const char ch : kForbidden) {
        if (token.find(ch) != std::string::npos) {
            throw SamError(SamErrorKind::Protocol,
                           "SAM " + std::string(field_name) +
                               " contains forbidden characters",
                           "");
        }
    }
    if (!allow_equals && token.find('=') != std::string::npos) {
        throw SamError(SamErrorKind::Protocol,
                       "SAM " + std::string(field_name) +
                           " contains forbidden characters",
                       "");
    }
    return token;
}

std::string validate_version(std::string_view value, std::string_view field_name) {
    const std::string token = validate_token(value, field_name, false);
    static const std::regex kPattern(R"(\d+\.\d+)");
    if (!std::regex_match(token, kPattern)) {
        throw SamError(SamErrorKind::Protocol,
                       "SAM " + std::string(field_name) + " version is invalid", "");
    }
    return token;
}

std::string validate_style(std::string_view value) {
    const std::string token = to_upper(validate_token(value, "STYLE", false));
    if (token != "STREAM" && token != "DATAGRAM" && token != "RAW") {
        throw SamError(SamErrorKind::Protocol, "SAM STYLE is invalid", "");
    }
    return token;
}

std::string validate_boolish(std::string_view value, std::string_view field_name) {
    std::string token = validate_token(value, field_name, false);
    std::transform(token.begin(), token.end(), token.begin(), [](unsigned char ch) {
        return static_cast<char>(std::tolower(ch));
    });
    if (token != "true" && token != "false") {
        throw SamError(SamErrorKind::Protocol,
                       "SAM " + std::string(field_name) + " is invalid", "");
    }
    return token;
}

std::vector<std::string> split_whitespace(std::string_view text) {
    std::vector<std::string> parts;
    std::istringstream stream{std::string(text)};
    std::string token;
    while (stream >> token) {
        parts.push_back(token);
    }
    return parts;
}

}  // namespace

bool is_sensitive_key(std::string_view key) {
    static const std::set<std::string> kSensitive{"PRIV", "PRIVATE", "DESTINATION",
                                                  "SIGNING_PRIVATE_KEY"};
    return kSensitive.count(to_upper(key)) > 0;
}

std::string redact_sam_line(std::string_view raw) {
    if (raw.empty()) {
        return std::string(raw);
    }
    std::vector<std::string> out;
    std::size_t start = 0;
    while (start <= raw.size()) {
        const std::size_t end = raw.find(' ', start);
        const std::string token(raw.substr(
            start, end == std::string_view::npos ? std::string_view::npos : end - start));
        const std::size_t eq = token.find('=');
        if (eq != std::string::npos) {
            const std::string key = token.substr(0, eq);
            const std::string value = token.substr(eq + 1);
            if (is_sensitive_key(key) && !value.empty()) {
                out.push_back(key + "=<redacted:" + std::to_string(value.size()) + "b>");
            } else {
                out.push_back(token);
            }
        } else {
            out.push_back(token);
        }
        if (end == std::string_view::npos) {
            break;
        }
        start = end + 1;
    }
    std::string joined;
    for (std::size_t i = 0; i < out.size(); ++i) {
        if (i > 0) {
            joined.push_back(' ');
        }
        joined += out[i];
    }
    return joined;
}

SamErrorKind map_result_to_error_kind(std::string_view result) {
    const std::string upper = to_upper(result);
    if (upper == "CANT_REACH_PEER") return SamErrorKind::CantReachPeer;
    if (upper == "DUPLICATED_DEST") return SamErrorKind::DuplicatedDest;
    if (upper == "DUPLICATED_ID") return SamErrorKind::DuplicatedId;
    if (upper == "I2P_ERROR") return SamErrorKind::I2pError;
    if (upper == "INVALID_ID") return SamErrorKind::InvalidId;
    if (upper == "INVALID_KEY") return SamErrorKind::InvalidKey;
    if (upper == "KEY_NOT_FOUND") return SamErrorKind::KeyNotFound;
    if (upper == "PEER_NOT_FOUND") return SamErrorKind::PeerNotFound;
    if (upper == "TIMEOUT") return SamErrorKind::Timeout;
    return SamErrorKind::Unknown;
}

std::optional<std::string> SamReply::field(std::string_view key) const {
    const auto it = fields.find(to_upper(key));
    if (it == fields.end()) {
        return std::nullopt;
    }
    return it->second;
}

std::string build_hello(std::string_view min_version, std::string_view max_version) {
    const std::string min_v = validate_version(min_version, "MIN");
    const std::string max_v = validate_version(max_version, "MAX");
    return "HELLO VERSION MIN=" + min_v + " MAX=" + max_v + "\n";
}

std::string build_dest_generate(int sig_type) {
    return "DEST GENERATE SIGNATURE_TYPE=" + std::to_string(sig_type) + "\n";
}

std::string build_naming_lookup(std::string_view name) {
    return "NAMING LOOKUP NAME=" + validate_token(name, "NAME", false) + "\n";
}

std::string build_session_create(
    std::string_view style, std::string_view session_id, std::string_view destination,
    const std::vector<std::pair<std::string, std::string>>& options,
    std::optional<int> sig_type) {
    // I2CP and streaming options are plain name=value tokens after DESTINATION;
    // SAM routers reject a standalone "OPTION" keyword.
    std::string option_string;
    if (sig_type.has_value()) {
        option_string = "SIGNATURE_TYPE=" + std::to_string(*sig_type);
    }
    for (const auto& [key, value] : options) {
        const std::string key_s = validate_token(key, "OPTION_KEY", false);
        const std::string val_s = validate_token(value, "OPTION_VALUE", false);
        if (!option_string.empty()) {
            option_string.push_back(' ');
        }
        option_string += key_s + "=" + val_s;
    }

    const std::string style_s = validate_style(style);
    const std::string session_s = validate_token(session_id, "ID", false);
    const std::string dest_s = validate_token(destination, "DESTINATION", true);

    // The space before the options is unconditional, matching the reference
    // implementation: with no options the line carries a trailing space.
    return "SESSION CREATE STYLE=" + style_s + " ID=" + session_s + " DESTINATION=" +
           dest_s + " " + option_string + "\n";
}

std::string build_stream_connect(std::string_view session_id,
                                 std::string_view destination,
                                 std::string_view silent) {
    const std::string sid = validate_token(session_id, "ID", false);
    const std::string dest = validate_token(destination, "DESTINATION", true);
    const std::string silent_flag = validate_boolish(silent, "SILENT");
    return "STREAM CONNECT ID=" + sid + " DESTINATION=" + dest + " SILENT=" +
           silent_flag + "\n";
}

std::string build_stream_accept(std::string_view session_id) {
    return "STREAM ACCEPT ID=" + validate_token(session_id, "ID", false) +
           " SILENT=false\n";
}

std::string build_stream_forward(std::string_view session_id, int port) {
    if (port < 1 || port > 65535) {
        throw SamError(SamErrorKind::Protocol, "SAM PORT is out of range", "");
    }
    // Trailing space before the newline is present in the reference builder.
    return "STREAM FORWARD ID=" + validate_token(session_id, "ID", false) + " PORT=" +
           std::to_string(port) + " \n";
}

SamReply parse_reply_line(std::string_view line) {
    const std::string raw = trim(line);
    if (raw.empty()) {
        throw SamError(SamErrorKind::Protocol, "Empty SAM reply", "");
    }
    const std::vector<std::string> parts = split_whitespace(raw);
    if (parts.size() < 2) {
        throw SamError(SamErrorKind::Protocol, "Malformed SAM reply",
                       redact_sam_line(raw));
    }

    SamReply reply;
    reply.command = to_upper(parts[0]);
    reply.topic = to_upper(parts[1]);
    for (std::size_t i = 2; i < parts.size(); ++i) {
        const std::size_t eq = parts[i].find('=');
        if (eq == std::string::npos) {
            continue;
        }
        reply.fields[to_upper(parts[i].substr(0, eq))] = parts[i].substr(eq + 1);
    }
    reply.raw_line = redact_sam_line(raw);
    return reply;
}

const SamReply& expect_ok(const SamReply& reply, std::string_view result_key) {
    const std::optional<std::string> raw_result = reply.field(result_key);
    const std::string result = raw_result.has_value() ? to_upper(*raw_result) : "";
    if (result == "OK") {
        return reply;
    }
    if (result.empty()) {
        // i2pd omits RESULT=OK on a successful DEST GENERATE and often on
        // SESSION CREATE; treat the payload-bearing replies as implicit OK.
        if (reply.command == "DEST" && reply.topic == "REPLY" &&
            reply.field("PUB").has_value() && reply.field("PRIV").has_value()) {
            return reply;
        }
        if (reply.command == "SESSION" && reply.topic == "STATUS" &&
            reply.field("DESTINATION").has_value()) {
            return reply;
        }
        throw SamError(SamErrorKind::Protocol, "SAM reply missing RESULT",
                       reply.raw_line);
    }
    const std::string message =
        reply.field("MESSAGE").value_or(reply.command + " " + reply.topic + " failed");
    throw SamError(map_result_to_error_kind(result), message, reply.raw_line);
}

}  // namespace i2pchat::sam
